### DISCUSSION TITLE
`interpreter.children` with spawned machines (in react)

### DISCUSSION THREAD

**Discussion by sunny-mittal** [statelyai/xstate#2137] [Q&A]
Hey all, 
I've had an issue I can quite identify when spawning machines via `invoke`. Unfortunately, I can't create a sandbox exhibiting the bug because it all works there so I'm hoping my actual application code will help with debugging. I have a parent "application" machine that spawns submachines when entering states that looks similar to the following (parts omitted due to proprietary restrictions as well as brevity):
```
const applicationMachine = Machine<Context, Schema, EventObject>(
  {
    id: "prequal",
    initial: "terms",
    context: DEFAULT_CONTEXT,
    states: {
      terms: {
        invoke: {
          id: "terms",
          src: termsMachine,
          onDone: {
            target: "info",
          },
        },
      },
...
```
and the `termsMachine` looks something like:
```
const termsMachine = Machine<Context, Schema, Event>({
  id: "terms",
  initial: "initializing",
  context: {
    authenticated: false,
  },
  states: {
    initializing: {
      always: [
        {
          target: "valid",
          cond: ({ authenticated }) => authenticated,
        },
        "idle",
      ],
    },
    idle: {...}
...
```
In my component, I pass the parent machine via React Context and try to access the spawned machine as follows:
```
  const [, , interpreter] = machine;

  const [{ context, matches }, send] = useService(
    interpreter.children.get("terms"),
  );
```
I get an error because `interpreter.children` is an empty Map but if I use the delay "hack" David mentions in https://github.com/davidkpiano/xstate/issues/795 and amend the parent machine to be:
```
const applicationMachine = Machine<Context, Schema, EventObject>(
  {
    id: "prequal",
    initial: "initializing",
    context: DEFAULT_CONTEXT,
    states: {
      initializing: {
        after: {
          1: "terms",
        },
      },
...
```
`children` is populated and everything works swimmingly. Am I missing something here? Thanks so much in advance!

**Comment by davidkpiano** [comment]
This part looks fragile:

```js
  const [, , interpreter] = machine;

  const [{ context, matches }, send] = useService(
    interpreter.children.get("terms"),
  );
```

What I would do is **pass the service through `props`**, which means that whatever is using that service should not even render unless the service actually exists:

```jsx
// in component
const termsService = interpreter.children.get('terms');

// ...

{termsService && <SomeComponent termsService={termsService} />}
```

And then in the component, you use it directly:

```jsx
const SomeComponent = ({ termsService }) => {
  const [...] = useService(termsService);
}
```

**Reply by sunny-mittal** [reply]
Oh good call...I like that...curious as to what looks fragile to you in my original, though.
